### ISSUE TITLE
Smartcard authentication doesn't work

### ISSUE THREAD

**Issue by ondrejholy** [FreeRDP/FreeRDP#8465]
One of our customers has [problems](https://bugzilla.redhat.com/show_bug.cgi?id=2105350) with Windows logon using IDEMIA Cosmo v8 PIV card. They use  `xfreerdp /u:<user> /d:<domain> /smartcard /v:<hostname> -sec-rdp` command line to connect to Window server 2012R2 from RHEL 8. Depending on FreeRDP version, the smartcard is not recognized at all (custom 2.8.1, or master builds), or is recognized but "No valid certificates" error is shown on the logon screen  (using the `freerdp-2.2.0-7.el8_5` package). The `/client-build-number:7601` and `2600` options don't help. This happens with the IDEMIA Cosmo v8 PIV card, but some other cards work with the same configuration, so this must be something related to that cards. There is a `freerdp-2.2.0-7.el8_5` [output](https://github.com/FreeRDP/FreeRDP/files/10113343/nosec-nla-v8.txt) generated over `/log-level:debug`. There are `SCARD_E_INSUFFICIENT_BUFFER` errors, though this might be just a red herring. We are trying to obtain also pscs debug log. Any idea what is going on, or how to further debug this issue?

**Comment by akallabeth** [comment]
good question, since you alredy mentioned other readers work.

Otherwise I´d split this in:
1. does it work with smartcard emulation builds? (logon et al)
2. does it work with some real hardware
3. what is the diff between the working and broken hardware (logs, ...)

**Comment by hardening** [comment]
Current master is temporary  broken, but once fixed, that could be worth give a try (as it supports smartcard logon and al).

**Comment by agilmore2** [comment]
I'm the original reporter. 
It's a card difference, not a reader problem.

Other cards work, other readers do not. 
What do you mean by "real" hardware?

How would I get notified when current master is ready for testing?

**Comment by akallabeth** [comment]
@agilmore2 ok, thank you for the update.
as for `master`, I´m working on https://github.com/FreeRDP/FreeRDP/pull/8461 which fixes a few bugs/regressions introduced with the latest pull requests.
once that is merged it will work again.

**Comment by ondrejholy** [comment]
@agilmore2, thanks for answering those questions here.

@akallabeth, there is [a log](https://github.com/FreeRDP/FreeRDP/files/10131055/sec-nla-oldcard-22.txt) for the old card from our customer case. One difference is that `SCARD_ATTR_DEVICE_FRIENDLY_NAME_W` and `SCARD_ATTR_ATR_STRING` attributes are requested for the new card, but not for the old card. Then there are those `SCARD_E_INSUFFICIENT_BUFFER` errors. Suspicious is that `SCardTransmit` failed with `SCARD_E_INSUFFICIENT_BUFFER` error, but `cbRecvLength: 65792` is the same for `Transmit_Call` and `Transmit_Return`, however, the expected size should be now in `cbRecvLength` as per the [pcsclite docs](https://pcsclite.apdu.fr/api/group__API.html#ga9a2d77242a271310269065e64633ab99). The promised `pcscd` debug log might give some explanation here...

```
[10:44:24:666] [12129:12173] [DEBUG][com.freerdp.channels.smartcard.client] - SCardTransmit (0x000900D0) FileId: 1 CompletionId: 0
[10:44:24:666] [12129:12173] [DEBUG][com.freerdp.channels.smartcard.client] - Transmit_Call {
[10:44:24:666] [12129:12173] [DEBUG][com.freerdp.channels.smartcard.client] - hContext: { 1179A51500000000 }
[10:44:24:666] [12129:12173] [DEBUG][com.freerdp.channels.smartcard.client] -   hContext: { 8C6B2E3800000000 }
[10:44:24:666] [12129:12173] [DEBUG][com.freerdp.channels.smartcard.client] - pioSendPci: dwProtocol: 2 cbExtraBytes: 0
[10:44:24:666] [12129:12173] [DEBUG][com.freerdp.channels.smartcard.client] - cbSendLength: 12
[10:44:24:666] [12129:12173] [DEBUG][com.freerdp.channels.smartcard.client] - pbSendBuffer: { 00CB3FFF0000035C017E0000 }
[10:44:24:666] [12129:12173] [DEBUG][com.freerdp.channels.smartcard.client] - pioRecvPci: null
[10:44:24:666] [12129:12173] [DEBUG][com.freerdp.channels.smartcard.client] - fpbRecvBufferIsNULL: 0 cbRecvLength: 65792
[10:44:24:666] [12129:12173] [DEBUG][com.freerdp.channels.smartcard.client] - }
[10:44:24:666] [12129:12206] [DEBUG][com.freerdp.channels.smartcard.client] - Transmit_Return {
[10:44:24:666] [12129:12206] [DEBUG][com.freerdp.channels.smartcard.client] -   ReturnCode: SCARD_E_INSUFFICIENT_BUFFER (0x80100008)
[10:44:24:666] [12129:12206] [DEBUG][com.freerdp.channels.smartcard.client] -   pioRecvPci: null
[10:44:24:666] [12129:12206] [DEBUG][com.freerdp.channels.smartcard.client] -   cbRecvLength: 65792
[10:44:24:666] [12129:12206] [DEBUG][com.freerdp.channels.smartcard.client] -   pbRecvBuffer: { 0000000000000000414720436172644D616E203331323120303020303000000000000000000000000000000000000000000000000000000000000000000000000000000000000000D1F7010000000000E01700DC5D7F00004B0045005900200041004700200043006100720064004D0061006E0020003300310032003100200030003000200030003000000000000000000000000000000081F70100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
[10:44:24:666] [12129:12206] [DEBUG][com.freerdp.channels.smartcard.client] - }
```

**Comment by hardening** [comment]
`SCARD_ATTR_DEVICE_FRIENDLY_NAME_W` and SCARD_E_INSUFFICIENT_BUFFER => that would match the last unicode changes

**Comment by akallabeth** [comment]
@ondrejholy ok, checking.

**Comment by akallabeth** [comment]
@ondrejholy as for the `SCARD_E_INSUFFICIENT_BUFFER` errors, do they occur only with `SCardTransmit` or other calls too?

**Comment by akallabeth** [comment]
@ondrejholy from the log and the code in `PCSC_SCardTransmit` only the `pcsc` library function is called and returns that error, so the `pcscd` log would really be interesting here.

**Comment by ondrejholy** [comment]
Just a note that those suspicious calls are probably caused by https://github.com/LudovicRousseau/PCSC/issues/137, the pcsc-lite build including the upstream fix has been provided to the customer and we are now waiting for feedback.

**Comment by WingsLikeEagles** [comment]
For FIPS compliance, seems like NLA connections with xfreerdp are still using NTLM, and not KERBEROS yet.  So NTLM uses algorythms that are not FIPS compliant.  This is stopping smartcard usage for any organizations that require FIPS compliance.

**Comment by akallabeth** [comment]
@WingsLikeEagles no, it uses kerberos if you have it available.
@ondrejholy this still an issue?

**Comment by ondrejholy** [comment]
The customer issue was closed on our side for inactivity, so let's do the same here as well...

**Comment by agilmore2** [comment]
Thanks for trying, I ran out of options to fix on the Linux side. Waiting
for a refresh of the Windows Server side to reattempt, and smart cards are
troublesome there as well. Unfortunately, Linux is leaving my
environment soon, and justifying my time on this one has been hard.
